modules/core/src/main/java/org/apache/ignite/internal/managers/encryption/GridEncryptionManager.java

```diff
@@ -308,23 +376,37 @@ public void onLocalJoin() {
         if (res != null)
             return res;
 
-        if (node.isClient())
+        if (isMasterKeyChangeInProgress()) {
+            // Prevents new nodes join to avoid inconsistency of the master key.
+            return new IgniteNodeValidationResult(ctx.localNodeId(),
+                "Master key change in progress! Node join is rejected. [node=" + node.id() + "]",
+                "Master key change in progress! Node join is rejected.");
+        }
+
+        if (node.isClient() || node.isDaemon())
             return null;
 
         res = validateNode(node);
 
         if (res != null)
             return res;
 
-        if (!discoData.hasJoiningNodeData()) {
-            U.quietAndInfo(log, "Joining node doesn't have encryption data [node=" + node.id() + "]");
+        NodeEncryptionKeys nodeEncKeys;
 
-            return null;
+        if (!discoData.hasJoiningNodeData() ||
+            (nodeEncKeys = (NodeEncryptionKeys)discoData.joiningNodeData()) == null) {
+            return new IgniteNodeValidationResult(ctx.localNodeId(),
+                "Joining node doesn't have encryption data [node=" + node.id() + "]",
+                "Joining node doesn't have encryption data.");
         }
 
-        NodeEncryptionKeys nodeEncKeys = (NodeEncryptionKeys)discoData.joiningNodeData();
+        if (!Arrays.equals(getSpi().masterKeyDigest(), nodeEncKeys.masterKeyDigest)) {
```

Comment:
We should check `masterKeyDigest` even joining node doesn't sent `nodeEncKeys`.
Consider the following case:

0. Node started.
1. `EncryptionSPI` configured. No encrypted caches created.
2. Node joins to the cluster with other encryption configuration.

Please, add explicit test to check this case.